{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6ccdbdd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.lines as mlines\n",
    "\n",
    "def parse_attr(attr, key):\n",
    "    for item in attr.split(';'):\n",
    "        if item.strip().startswith(key):\n",
    "            return item.strip().split(' ')[1].replace('\"', '')\n",
    "    return None\n",
    "\n",
    "def base_counts_to_df(base_counts, total_counts, region_start, region_end):\n",
    "    positions = range(region_start, region_end + 1)\n",
    "    records = []\n",
    "    for pos in positions:\n",
    "        row = {'pos': pos}\n",
    "        for base in ['A', 'T', 'C', 'G']:\n",
    "            row[f'{base}_count'] = base_counts[pos].get(base, 0)\n",
    "        row['total_count'] = total_counts.get(pos, 0)\n",
    "        records.append(row)\n",
    "    return pd.DataFrame(records)\n",
    "\n",
    "def plot_modifications(\n",
    "    bamfile, gtf_file, bed_file,modcode,\n",
    "    region_chr, region_start, region_end, region_strand,\n",
    "    output_figure_path\n",
    "):\n",
    "\n",
    "    base_counts = defaultdict(lambda: defaultdict(int))\n",
    "    total_counts = defaultdict(int)\n",
    "    \n",
    "    print('Counting Modified and Unmodified Counts')\n",
    "\n",
    "    bam = pysam.AlignmentFile(bamfile, \"rb\")\n",
    "    \n",
    "    for pileupcolumn in bam.pileup(region_chr, region_start, region_end, truncate=True):\n",
    "        pos = pileupcolumn.reference_pos\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            read = pileupread.alignment\n",
    "            if read.is_supplementary or read.is_secondary:\n",
    "                continue\n",
    "            strand = '-' if read.is_reverse else '+'\n",
    "            if strand != region_strand:\n",
    "                continue\n",
    "            if not pileupread.is_del and not pileupread.is_refskip:\n",
    "                base = read.query_sequence[pileupread.query_position]\n",
    "                base_counts[pos][base] += 1\n",
    "\n",
    "    for read in bam.fetch(region_chr, region_start, region_end):\n",
    "        if read.is_secondary: #read.is_supplementary or \n",
    "            continue\n",
    "        strand = '-' if read.is_reverse else '+'\n",
    "        if strand != region_strand:\n",
    "            continue\n",
    "        ref_positions = read.get_reference_positions(full_length=False)\n",
    "        for ref_pos in ref_positions:\n",
    "            if ref_pos is None:\n",
    "                continue\n",
    "            if region_start <= ref_pos <= region_end:\n",
    "                total_counts[ref_pos] += 1\n",
    "\n",
    "    bam.close()\n",
    "\n",
    "    coverage_df = base_counts_to_df(base_counts, total_counts, region_start, region_end)\n",
    "\n",
    "    bed_cols = [\n",
    "        'chrom', 'start', 'end', 'mod', 'score', 'strand',\n",
    "        'thickStart', 'thickEnd', 'itemRgb',\n",
    "        'total_call', 'mod_ratio', 'modified_call'\n",
    "    ]\n",
    "    df_bed = pd.read_csv(\n",
    "        bed_file,\n",
    "        sep='\\t',\n",
    "        names=bed_cols,\n",
    "        header=None,\n",
    "        usecols=range(12)\n",
    "    )\n",
    "    df_bed['pos'] = df_bed['end']\n",
    "\n",
    "    region_bed = df_bed[\n",
    "        (df_bed['chrom'] == region_chr) &\n",
    "        (df_bed['pos'] >= region_start) &\n",
    "        (df_bed['pos'] <= region_end) &\n",
    "        (df_bed['strand'] == region_strand)\n",
    "    ]\n",
    "\n",
    "    modification_counts = (\n",
    "        region_bed.groupby(['pos', 'mod'])['modified_call']\n",
    "        .sum()\n",
    "        .unstack(fill_value=0)\n",
    "        .sort_index()\n",
    "    )\n",
    "\n",
    "    modification_counts.reset_index(inplace=True)\n",
    "\n",
    "    merged_df = pd.merge(coverage_df, modification_counts, on='pos', how='left')\n",
    "    merged_df = merged_df.fillna(0)\n",
    "\n",
    "    print('Extracting gene annotations')\n",
    "\n",
    "    gtf_cols = [\"chrom\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "    gtf = pd.read_csv(gtf_file, sep='\\t', comment='#', header=None, names=gtf_cols)\n",
    "\n",
    "    gtf['gene_name'] = gtf['attribute'].apply(lambda x: parse_attr(x, 'gene_name'))\n",
    "    gtf['gene_id'] = gtf['attribute'].apply(lambda x: parse_attr(x, 'gene_id'))\n",
    "    gtf['transcript_id'] = gtf['attribute'].apply(lambda x: parse_attr(x, 'transcript_id'))\n",
    "\n",
    "    exons = gtf[\n",
    "        (gtf['feature'] == 'exon') &\n",
    "        (gtf['chrom'] == region_chr) &\n",
    "        (gtf['start'] <= region_end) &\n",
    "        (gtf['end'] >= region_start) &\n",
    "        (gtf['strand'] == region_strand)\n",
    "    ]\n",
    "\n",
    "    transcripts = exons.groupby('transcript_id')\n",
    "\n",
    "    positions = merged_df['pos']\n",
    "    base_total = merged_df['total_count']\n",
    "    A_total = merged_df['A_count']\n",
    "\n",
    "    a_total = merged_df.get( modcode, pd.Series(0, index=positions))\n",
    "    \n",
    "    print('Start drawing')\n",
    "\n",
    "    fig, (ax1, ax2) = plt.subplots(\n",
    "        2, 1, figsize=(20, 8), sharex=True,\n",
    "        gridspec_kw={'height_ratios': [3, 1]}\n",
    "    )\n",
    "\n",
    "    bar_width = 1\n",
    "    \n",
    "    ax1.plot(positions, base_total, color='lightgrey', label='Total base count')\n",
    "    ax1.fill_between(positions, 0, base_total, color='lightgrey', alpha=0.3)\n",
    "\n",
    "    \n",
    "#    ax1.bar(positions, A_total, width=bar_width, color='lightblue', label='A count')\n",
    "#    ax1.bar(positions, a_total, width=bar_width, color='red', label='m6A count')\n",
    "    \n",
    "    nonzero_mask = a_total > 0\n",
    "    markerline, stemlines, baseline = ax1.stem(\n",
    "        positions[nonzero_mask],\n",
    "        a_total[nonzero_mask],\n",
    "        basefmt=' ')\n",
    "    markerline.set_color('#ffb178')  # 改 marker 颜色\n",
    "    stemlines.set_color('#ffb178')   # 改 stem 线颜色\n",
    "    plt.setp(markerline, markersize=2)\n",
    "    plt.setp(stemlines, linewidth=1)\n",
    "\n",
    "\n",
    "    ax1.set_ylabel('Counts')\n",
    "    ax1.legend()\n",
    "\n",
    "    ax1.tick_params(axis='x', labelbottom=True)\n",
    "\n",
    "    strand_color = {'+': 'skyblue', '-': 'salmon'}\n",
    "    intron_line_color = 'black'\n",
    "    y_step = 1\n",
    "    y_base = 0\n",
    "\n",
    "    for idx, (transcript_id, group) in enumerate(transcripts):\n",
    "        strand = group['strand'].iloc[0]\n",
    "        gene_name = group['gene_name'].iloc[0] or transcript_id\n",
    "        y = y_base + idx * y_step\n",
    "\n",
    "        exon_starts = []\n",
    "        exon_ends = []\n",
    "        for _, row in group.iterrows():\n",
    "            exon_start = row['start']\n",
    "            exon_end = row['end']\n",
    "            exon_starts.append(exon_start)\n",
    "            exon_ends.append(exon_end)\n",
    "\n",
    "            ax2.add_patch(mpatches.Rectangle(\n",
    "                (exon_start, y),\n",
    "                exon_end - exon_start,\n",
    "                0.4,\n",
    "                facecolor=strand_color.get(strand, 'grey'),\n",
    "                edgecolor='black'\n",
    "            ))\n",
    "\n",
    "        exon_centers = [(start + end) / 2 for start, end in zip(exon_starts, exon_ends)]\n",
    "        for i in range(len(exon_centers) - 1):\n",
    "            ax2.add_line(mlines.Line2D(\n",
    "                [exon_ends[i], exon_starts[i+1]],\n",
    "                [y + 0.2, y + 0.2],\n",
    "                color=intron_line_color,\n",
    "                linewidth=0.5\n",
    "            ))\n",
    "\n",
    "        region_center = (region_start + region_end) / 2\n",
    "        label_offset = 0.1\n",
    "        ax2.text(region_center, y + label_offset, gene_name, ha='center', fontsize=8)\n",
    "\n",
    "    ax2.set_ylim(-0.5, len(transcripts) * y_step)\n",
    "\n",
    "    ax2.axis('off')\n",
    "\n",
    "    ax2.set_xlabel('Genomic Position')\n",
    "\n",
    "    region_xlim = (region_start, region_end)\n",
    "    ax1.set_xlim(region_xlim)\n",
    "    ax2.set_xlim(region_xlim)\n",
    "\n",
    "    fig.subplots_adjust(left=0.05, right=0.95, top=0.95, bottom=0.15)\n",
    "\n",
    "    plt.savefig(output_figure_path, dpi=300, bbox_inches='tight')\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c5cd7dfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counting Modified and Unmodified Counts\n",
      "Extracting gene annotations\n",
      "Start drawing\n"
     ]
    }
   ],
   "source": [
    "plot_modifications(\n",
    "    bamfile=\"/scratch/lb4489/project/dRNA/HIP_mum_mapped.sorted.bam\",\n",
    "    gtf_file=\"/scratch/lb4489/bioindex/gencode.vM33.annotation.gtf\",\n",
    "    bed_file=\"/scratch/lb4489/project/dRNA/modkit_all_filted.bed\",\n",
    "    region_chr=\"chr18\",\n",
    "    region_start=61058690,\n",
    "    region_end=61121224,\n",
    "    region_strand='+',\n",
    "    output_figure_path=\"/scratch/lb4489/project/dRNA/Camk2a_figure.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3dbf6c9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counting Modified and Unmodified Counts\n",
      "Extracting gene annotations\n",
      "Start drawing\n"
     ]
    }
   ],
   "source": [
    "plot_modifications(\n",
    "    bamfile=\"/scratch/lb4489/project/dRNA/HIP_mum_mapped.sorted.bam\",\n",
    "    gtf_file=\"/scratch/lb4489/bioindex/gencode.vM33.annotation.gtf\",\n",
    "    bed_file=\"/scratch/lb4489/project/dRNA/modkit_all_filted.bed\",\n",
    "    region_chr=\"chr18\",\n",
    "    region_start=61115649,\n",
    "    region_end=61122596,\n",
    "    region_strand='+',\n",
    "    output_figure_path=\"/scratch/lb4489/project/dRNA/Camk2a_lastexon_figure.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0adfb95d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counting Modified and Unmodified Counts\n",
      "Extracting gene annotations\n",
      "Start drawing\n"
     ]
    }
   ],
   "source": [
    "plot_modifications(\n",
    "    bamfile=\"/scratch/lb4489/project/dRNA/HIP_mum_mapped.sorted.bam\",\n",
    "    gtf_file=\"/scratch/lb4489/bioindex/gencode.vM33.annotation.gtf\",\n",
    "    bed_file=\"/scratch/lb4489/project/dRNA/modkit_all_filted.bed\",\n",
    "    region_chr=\"chr11\",\n",
    "    region_start=5919644,\n",
    "    region_end=6015748,\n",
    "    region_strand='-',\n",
    "    output_figure_path=\"/scratch/lb4489/project/dRNA/Camk2b_figure.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "20dbd5e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counting Modified and Unmodified Counts\n",
      "Extracting gene annotations\n",
      "Start drawing\n"
     ]
    }
   ],
   "source": [
    "plot_modifications(\n",
    "    bamfile=\"/scratch/lb4489/project/dRNA/HIP_mum_mapped.sorted.bam\",\n",
    "    gtf_file=\"/scratch/lb4489/bioindex/gencode.vM33.annotation.gtf\",\n",
    "    bed_file=\"/scratch/lb4489/project/dRNA/modkit_all_filted.bed\",\n",
    "    region_chr=\"chr11\",\n",
    "    modcode='a',\n",
    "    region_start=5919080,\n",
    "    region_end=5922638,\n",
    "    region_strand='-',\n",
    "    output_figure_path=\"/scratch/lb4489/project/dRNA/Camk2b_lastexon_figure.png\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e0c8f651",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counting Modified and Unmodified Counts\n",
      "Extracting gene annotations\n",
      "Start drawing\n"
     ]
    }
   ],
   "source": [
    "plot_modifications(\n",
    "    bamfile=\"/scratch/lb4489/project/dRNA/HIP_mum_mapped.sorted.bam\",\n",
    "    gtf_file=\"/scratch/lb4489/bioindex/gencode.vM33.annotation.gtf\",\n",
    "    bed_file=\"/scratch/lb4489/project/dRNA/modkit_all_filted.bed\",\n",
    "    region_chr=\"chr11\",\n",
    "    modcode='m',\n",
    "    region_start=5919080,\n",
    "    region_end=5922638,\n",
    "    region_strand='-',\n",
    "    output_figure_path=\"/scratch/lb4489/project/dRNA/Camk2b_lastexon_figure_m5c.png\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c82c4b9a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
